app/services/lms_sync_service.py

```diff
@@ -0,0 +1,241 @@
+            for instructor in db_instructors:
+                instructor_pid = await self.canvas_service.get_pid_from_onyen(instructor.onyen)
+                if instructor_pid not in [i['sis_user_id'] for i in canvas_instructors]:
+                    await self.canvas_service.unassociate_pid_from_user(instructor.onyen)
+                    await self.instructor_service.delete_user(instructor.onyen)
+        
+        for instructor in canvas_instructors:
+            pid = instructor['sis_user_id']
+            user_info = self.ldap_service.get_user_info(pid)
+
+            try:
+                await self.instructor_service.get_user_by_onyen(user_info.onyen)
+
+            except UserNotFoundException:
+                #create a new instructor
+                await self.instructor_service.create_instructor(
+                    onyen=user_info.onyen,
+                    name=instructor['name'],
+                    email=instructor['email']
+                )
+                await self.canvas_service.associate_pid_to_user(user_info.onyen, pid)
+
+        return canvas_instructors
+
+    async def upload_grades_from_csv(self, grade_csv):
+        df = pd.read_csv(io.StringIO(grade_csv.decode("utf-8")))
+        
+        for index, row in df.iterrows():
+            try:
+                assignment_id = self._extract_id(row['file'])
+                user_pid = await self.canvas_service.get_pid_from_onyen(row['onyen'])
+                students = await self.canvas_service.get_users({
+                    "enrollment_type": "student"
+                })
+                student = next((student for student in students if student['sis_user_id'] == user_pid), None)
+
+                await self.canvas_service.upload_grades(assignment_id, student["id"], row['percent_correct'])
+
+            except Exception as e:
+                raise HTTPException(status_code=500, detail=str(e))
+            
+        return {
+            "message": "Successfully uploaded grades"
+        }
+
+    async def upsync_assignment(self, assignment):
+        unlock_date = assignment.available_date.strftime("%Y-%m-%dT%H:%M:%S")
+        due_date = assignment.due_date.strftime("%Y-%m-%dT%H:%M:%S")
+        try:
+            await self.canvas_service.update_assignment(assignment.id,
+            {
+                "assignment": {
+                    "unlock_at": unlock_date,
+                    "due_at": due_date
+                }
+            })
+        
+        except Exception as e:
+            raise HTTPException(status_code=500, detail=str(e))
+        
+        return {
+            "message": "Successfully updated assignment"
+        }
+
+
+    async def downsync(self):
+        await asyncio.gather(
+            self.sync_course(),
+            self.sync_assignments(),
+            self.sync_students(),
+            self.sync_instructors()
+        )
+
+        return {
+            "message": "Successfully synced the LMS with the database"
```

Comment:
Weird format to return, I'd prefer it just to be void

Reply:
done